### ISSUE TITLE
`showyourwork build` not compatible with Python 3.11

### ISSUE THREAD

**Issue by rwegener2** [showyourwork/showyourwork#258]
I was following the [Quickstart docs](https://show-your.work/en/latest/quickstart/) recently when I got the following error:
The command `showyourwork build` resulted in
```
Creating conda environment /Users/rwegener/.pyenv/versions/3.11.1/lib/python3.11/site-packages/showyourwork/workflow/envs/environment.yml...
Downloading and installing remote packages.
Environment for /Users/rwegener/.pyenv/versions/3.11.1/lib/python3.11/site-packages/showyourwork/workflow/envs/environment.yml created (location: .snakemake/conda/64e7186f4fef8f7627c97a4c3040e45a_)
Traceback (most recent call last):
  File "/Users/rwegener/.pyenv/versions/3.11.1/lib/python3.11/site-packages/snakemake/__init__.py", line 733, in snakemake
    success = workflow.execute(
              ^^^^^^^^^^^^^^^^^
  File "/Users/rwegener/.pyenv/versions/3.11.1/lib/python3.11/site-packages/snakemake/workflow.py", line 946, in execute
    self.scheduler = JobScheduler(
                     ^^^^^^^^^^^^^
  File "/Users/rwegener/.pyenv/versions/3.11.1/lib/python3.11/site-packages/snakemake/scheduler.py", line 108, in __init__
    from ratelimiter import RateLimiter
  File "/Users/rwegener/.pyenv/versions/3.11.1/lib/python3.11/site-packages/ratelimiter.py", line 36, in <module>
    class RateLimiter(object):
  File "/Users/rwegener/.pyenv/versions/3.11.1/lib/python3.11/site-packages/ratelimiter.py", line 127, in RateLimiter
    __aexit__ = asyncio.coroutine(__exit__)
                ^^^^^^^^^^^^^^^^^
AttributeError: module 'asyncio' has no attribute 'coroutine'
```

It seems from some googling ([this Stack Overflow post](https://stackoverflow.com/questions/74345065/attributeerror-module-asyncio-has-no-attribute-coroutine-in-python-3-11)) like some syntax in the `asyncio` module was depreciated with Python 3.11. It doesn't seem that sytnax is used in the `showyourwork` repository, but perhaps it is called in `snakemake`.

I was able to avoid the issue by installing Python 10 and starting over with the `showyourwork` setup. I still wanted to flag the issue in Github.

Thanks all for making such a great package - I'm excited to get into it!

**Comment by dfm** [comment]
Thanks for the report! We [pin the version of `snakemake`](https://github.com/showyourwork/showyourwork/blob/f75a063e2f4a81c899924d6d9121a465e9bfd9ec/setup.py#L29) since we use some pretty brittle patches to get all the behavior we want so even if `snakemake` is compatible with 3.11 we're probably lagging behind. For now, the best bet is probably to use the older Python (we've mostly been using 3.9, doh!) and let's try to get this updated soon!

**Comment by HealthyPear** [comment]
Hi,

I got the same error.

I solved it by performing an additional step: I defined a _showyourwork_ conda env like this,

```yaml
name: showyourwork
channels:
  - conda-forge
dependencies:
  - python<3.11
  - pip
  - pip:
      - showyourwork
```

but then I do `showyourwork build` from within my new `showyourwork` conda environment, so I don't know what is the consequence of this for the _internal_ conda environement.

Related to this, I noticed that this project requires conda, but it's only packaged with pip which is a bit weird to see - any chance you'll make a conda package?

**Comment by dfm** [comment]
Yeah - this issue is still open because the currently released SYW still isn't compatible with Python 3.11. Working on it! :D 

> Related to this, I noticed that this project requires conda, but it's only packaged with pip which is a bit weird to see - any chance you'll make a conda package?

There's actually a somewhat long story here, but the issue comes from the fact that our dependency `snakemake` is only on `biconda`, not `conda-forge` and the `snakemake` developers don't want to change that. Last time we looked into it, `conda-forge` needed all dependencies to be hosted on `conda-forge`, so we were hosed. An alternative would be to try to submit it to `bioconda`, but I'm less excited about that.

**Comment by HealthyPear** [comment]
> Yeah - this issue is still open because the currently released SYW still isn't compatible with Python 3.11. Working on it! :D

Since this package is installed with _setuptools_, it is easy to temporarily "fix" this bug by specifying

`python_requires=">=3.8,<3.11",`

in your `setup.py` and update the documentation accordingly.

>=0.5.0,< 0.7.0'

> There's actually a somewhat long story here, but the issue comes from the fact that our dependency snakemake is only on biconda, not conda-forge and the snakemake developers don't want to change that. Last time we looked into it, conda-forge needed all dependencies to be hosted on conda-forge, so we were hosed. An alternative would be to try to submit it to bioconda, but I'm less excited about that.

I see. I didn't know about this complication.

-----

It's not really on-topic, but I think the docs lack an explanation of how the internal conda environment works.
I don't understand if it gets created at every build and how to activate it or interact with it (if it's even expected to be possible from this package).

If you prefer I can open a discussion (not really an issue in this case).

**Comment by dfm** [comment]
> It's not really on-topic, but I think the docs lack an explanation of how the internal conda environment works.
I don't understand if it gets created at every build and how to activate it or interact with it (if it's even expected to be possible from this package).

@rwegener2 recently contributed more docs discussing this (https://github.com/showyourwork/showyourwork/pull/309), but it's still pretty minimal. I think it makes sense to open as an issue if you don't find the docs sufficient - please do!
